### DISCUSSION TITLE
Typst syntax: Unpack Array

### DISCUSSION THREAD

**Discussion by davidleejy** [EpicEricEE/typst-plugins#5] [Q&A] [answered]
Hi Eric, thanks for assisting me with a question I had about defining custom shorthands in a Q&A I had posted in the discussion board on Typst's repository earlier today.

Would you happen to know how to unpack an array so that its elements can be used as arguments in `#show: shorthands.with(...)`. The reason for doing so is shared in this annotated screenshot:

![Screenshot 2024-02-18 at 6 58 07 PM](https://github.com/EpicEricEE/typst-plugins/assets/1690072/8b1f2e5a-756d-490b-aed9-2b5ab25bb9c0)

**EDIT**: 

1. I just realized this wouldn't directly work because the the `#show: shorthands.with(...)` has to be executed before the lines of text demonstrating what has been defined as shorthands. However, with some slight modification, I believe I'd still be able to create a preview that shows the active custom shorthands 😄.

2. Alternatively, it would be nice if the `quickmath` package had a function to print all custom shorthands that have been defined. Would be extra nice if shorthands could be paired with a user-defined comment e.g.

```typst
#import "@preview/quick-maths:0.1.0": shorthands

#show: shorthands.with(
  ($>->>$, $⤖$, "bijective arrow right"), 
  ($<-<<$, $⬻$, "bijective arrow left"), 
  ($<-x-$, $⬾$, "crossed arrow left"), 
  ($+-$, $plus.minus$, "+/- symbol"),
  ($<=$, math.arrow.l.double, "backward implication arrow. Warning: overwrites default shorthand for ≤")
)

#shorthands.print_all()   // Prints all custom shorthands defined.
```

Cheers, and thanks for developing and releasing quickmath 😃

**Comment by EpicEricEE** [comment, marked as answer]
The syntax for unpacking in typst is `shorthands.with(..custom_shorthands)`.

Also, you can probably use "scoped" shorthands like this:
```typ
#let custom_shorthands = ()

#custom_shorthands.push(...)
#shorthands(..custom_shorthands)[
  Arrows: \
  ...
]

#custom_shorthands.push(...)
...

#show: shorthands.with(..custom_shorthands)
```

**Reply by davidleejy** [reply]
Thanks for your help with unpacking an array in Typst. 

Pardon my unfamiliarity with Typst syntax, but what does "scoped" shorthands mean? I'm guessing it's the snippet:

```Typst
#shorthands(custom_shorthands)[
  Arrows: \
  ...
]
```

It looks like a key-value mapping, but I'm not sure I see how to use it in a clever manner 😓

**Reply by EpicEricEE** [reply]
`shorthands` is a function that takes a list and some content and applies the shorthands to that content. When you use it with a show rule, it will apply the shorthands to everything after the show rule. So if you want the shorthands in your first code snippet to work before the show rule, you can do it like this:
```typ
#let custom_shorthands = ()

// Bijective arrows:
#custom_shorthands.push(($>->>$, $⤖$))
#custom_shorthands.push(($<-<<$, $⬻$))
#shorthands(..custom_shorthands)[
  Bijective arrows: \
  >->> \<-<< gives $>->> <-<<$
]

// Plus minus:
#custom_shorthands.push(($+-$, $plus.minus$))
#shorthands(..custom_shorthands)[
  Plus minus: \
  +- gives $+-$
]

// <= arrow:
#custom_shorthands.push(($<=$, math.arrow.l.double))
#shorthands(..custom_shorthands)[
  Arrow: \
  <= gives $<=$ ($a <= b$) // Doesn't work with $<=$ alone
]

#show: shorthands.with(..custom_shorthands)
```

A `print-all` function would be hard to implement nicely, as the shorthand itself isn't stored in the original form. Typst already has the inbuilt shorthand `$>->$`, so `$>->>$` is really stored as the sequence `(↣, >)`.

**Reply by davidleejy** [reply]
Thanks once again, and am happy to report that your solution works 😀 .
